File: backend/app/schemas/glossary_validation.py
"""Public Phase 9 glossary validation job, result, and match contracts."""

from __future__ import annotations

from datetime import datetime
from typing import Any
from uuid import UUID

from pydantic import Field, field_validator

from app.models.glossary_enums import (
    GlossaryLanguageCode,
    GlossaryMatchType,
    GlossarySourceType,
    GlossaryValidationJobType,
    GlossaryValidationStatus,
)
from app.schemas.base import ApiSchema
from app.schemas.common import PaginationData


class GlossaryValidationRequest(ApiSchema):
    document_file_id: UUID
    compliance_run_id: UUID | None = None
    profile_ids: list[UUID] = Field(default_factory=list, max_length=100)
    force: bool = False

    @field_validator("profile_ids")
    @classmethod
    def deduplicate_profiles(cls, value: list[UUID]) -> list[UUID]:
        return list(dict.fromkeys(value))


class GlossaryRevalidationRequest(ApiSchema):
    reason: str = Field(min_length=1, max_length=2000)
    profile_ids: list[UUID] = Field(default_factory=list, max_length=100)

    @field_validator("reason", mode="before")
    @classmethod
    def reason_is_required(cls, value: str) -> str:
        normalized = value.strip()
        if not normalized:
            raise ValueError("Reason must not be empty.")
        return normalized

    @field_validator("profile_ids")
    @classmethod
    def deduplicate_profiles(cls, value: list[UUID]) -> list[UUID]:
        return list(dict.fromkeys(value))


class GlossaryValidationQueuedResponse(ApiSchema):
    job_id: UUID
    run_id: UUID
    status: GlossaryValidationStatus
    progress: int = Field(ge=0, le=100)
    document_file_id: UUID
    reused_existing_result: bool = False


class GlossaryValidationRunResponse(ApiSchema):
    id: UUID
    job_id: UUID
    document_id: UUID
    document_revision_id: UUID
    document_file_id: UUID
    compliance_run_id: UUID | None
    language_detection_run_id: UUID
    glossary_profile_ids: list[UUID]
    profile_snapshots: list[dict[str, Any]]
    job_type: GlossaryValidationJobType
    status: GlossaryValidationStatus
    progress: int = Field(ge=0, le=100)
    current_stage: str | None
    source_content_hash: str
    total_terms: int = Field(ge=0)
    matched_terms: int = Field(ge=0)
    preferred_term_matches: int = Field(ge=0)
    forbidden_term_matches: int = Field(ge=0)
    missing_required_translations: int = Field(ge=0)
    inconsistent_terms: int = Field(ge=0)
    exception_applied_count: int = Field(ge=0)
    total_findings: int = Field(ge=0)
    metrics: dict[str, Any]
    warnings: list[str] | list[dict[str, Any]]
    error_code: str | None
    error_message: str | None
    requested_by: UUID | None
    requested_at: datetime
    started_at: datetime | None
    completed_at: datetime | None
    failed_at: datetime | None
    cancel_requested_at: datetime | None
    cancelled_at: datetime | None
    created_at: datetime
    updated_at: datetime


class GlossaryValidationJobListResponse(
    PaginationData[GlossaryValidationRunResponse]
):
    pass


class GlossaryValidationSummaryResponse(ApiSchema):
    run_id: UUID
    status: GlossaryValidationStatus
    total_terms: int = Field(ge=0)
    matched_terms: int = Field(ge=0)
    preferred_term_matches: int = Field(ge=0)
    forbidden_term_matches: int = Field(ge=0)
    missing_required_translations: int = Field(ge=0)
    inconsistent_terms: int = Field(ge=0)
    exception_applied_count: int = Field(ge=0)
    total_findings: int = Field(ge=0)
    match_count: int = Field(ge=0)
    language_counts: dict[str, int] = Field(default_factory=dict)
    finding_counts: dict[str, int] = Field(default_factory=dict)
    metrics: dict[str, Any] = Field(default_factory=dict)
    warnings: list[str] | list[dict[str, Any]] = Field(default_factory=list)


class GlossaryMatchResponse(ApiSchema):
    id: UUID
    glossary_validation_run_id: UUID
    glossary_term_id: UUID
    glossary_translation_id: UUID | None
    glossary_variant_id: UUID | None
    term_code: str | None = None
    concept_name: str | None = None
    language_code: GlossaryLanguageCode
    source_type: GlossarySourceType
    extracted_block_id: UUID | None
    ocr_block_id: UUID | None
    container_id: UUID | None
    detected_section_id: UUID | None
    source_reference: str
    matched_text: str
    normalised_matched_text: str
    start_offset: int = Field(ge=0)
    end_offset: int = Field(ge=0)
    match_type: GlossaryMatchType
    is_preferred: bool
    is_forbidden: bool
    exception_id: UUID | None
    metadata: dict[str, Any]
    created_at: datetime


class GlossaryMatchListResponse(PaginationData[GlossaryMatchResponse]):
    pass


class GlossaryValidationHistoryResponse(
    PaginationData[GlossaryValidationRunResponse]
):
    pass


class GlossaryFindingSignal(ApiSchema):
    """Finding data returned before/alongside Phase 8 workflow persistence."""

    id: UUID
    finding_code: str
    severity: str
    status: str
    title: str
    description: str
    recommendation: str
    glossary_term_id: UUID
    language_code: GlossaryLanguageCode | None = None
    source_reference: str | None = None
    extracted_block_id: UUID | None = None
    ocr_block_id: UUID | None = None
    translation_group_id: UUID | None = None
    exception_id: UUID | None = None
    metrics: dict[str, Any] = Field(default_factory=dict)
    is_repeat: bool = False
    previous_finding_id: UUID | None = None
    created_at: datetime


class GlossaryFindingListResponse(PaginationData[GlossaryFindingSignal]):
    pass

File: frontend/src/types/glossaryValidation.ts
import type { GlossaryLanguageCode, GlossarySeverity } from './glossary';
import type { PaginatedData } from './masterData';

export const glossaryValidationJobStatuses = [
  'QUEUED',
  'LOADING_CONTEXT',
  'MATCHING_TERMS',
  'VALIDATING_TERMS',
  'GENERATING_FINDINGS',
  'PERSISTING',
  'COMPLETED',
  'PARTIALLY_COMPLETED',
  'FAILED',
  'CANCEL_REQUESTED',
  'CANCELLED',
] as const;

export type GlossaryValidationJobStatus =
  (typeof glossaryValidationJobStatuses)[number];
export type GlossaryValidationRunStatus =
  'COMPLETED' | 'PARTIALLY_COMPLETED' | 'FAILED';
export type GlossaryValidationJobType = 'INITIAL' | 'REVALIDATION' | 'MANUAL';

export const terminalGlossaryValidationJobStatuses: readonly GlossaryValidationJobStatus[] =
  ['COMPLETED', 'PARTIALLY_COMPLETED', 'FAILED', 'CANCELLED'];

export const isTerminalGlossaryValidationJobStatus = (
  status: GlossaryValidationJobStatus,
): boolean => terminalGlossaryValidationJobStatuses.includes(status);

export interface GlossaryValidationRequest {
  documentFileId: string;
  complianceRunId?: string | null;
  profileIds?: readonly string[];
  force?: boolean;
}

export interface GlossaryValidationQueuedResult {
  jobId: string;
  runId: string;
  status: GlossaryValidationJobStatus;
  progress: number;
  documentFileId: string;
  reusedExistingResult: boolean;
}

export interface GlossaryValidationJobListParams {
  page: number;
  pageSize: number;
  status?: GlossaryValidationJobStatus | readonly GlossaryValidationJobStatus[];
  departmentId?: string;
  search?: string;
}

export interface GlossaryValidationRun {
  id: string;
  jobId: string;
  documentId: string;
  documentRevisionId: string;
  documentFileId: string;
  complianceRunId: string | null;
  languageDetectionRunId: string;
  glossaryProfileIds: readonly string[];
  profileSnapshots: readonly Readonly<Record<string, unknown>>[];
  jobType: GlossaryValidationJobType;
  status: GlossaryValidationJobStatus;
  progress: number;
  currentStage: string | null;
  sourceContentHash: string;
  totalTerms: number;
  matchedTerms: number;
  preferredTermMatches: number;
  forbiddenTermMatches: number;
  missingRequiredTranslations: number;
  inconsistentTerms: number;
  exceptionAppliedCount: number;
  totalFindings: number;
  metrics: Readonly<Record<string, unknown>>;
  warnings: readonly (string | Readonly<Record<string, unknown>>)[];
  errorCode: string | null;
  errorMessage: string | null;
  requestedBy: string | null;
  requestedAt: string;
  startedAt: string | null;
  completedAt: string | null;
  failedAt: string | null;
  cancelRequestedAt: string | null;
  cancelledAt: string | null;
  createdAt: string;
  updatedAt: string;
}

export type GlossaryValidationJob = GlossaryValidationRun;
export type GlossaryValidationJobList = PaginatedData<GlossaryValidationJob>;

export interface GlossaryValidationSummary {
  runId: string;
  status: GlossaryValidationJobStatus;
  totalTerms: number;
  matchedTerms: number;
  preferredTermMatches: number;
  forbiddenTermMatches: number;
  missingRequiredTranslations: number;
  inconsistentTerms: number;
  exceptionAppliedCount: number;
  totalFindings: number;
  matchCount: number;
  languageCounts: Readonly<Record<string, number>>;
  findingCounts: Readonly<Record<string, number>>;
  metrics: Readonly<Record<string, unknown>>;
  warnings: readonly (string | Readonly<Record<string, unknown>>)[];
  termsEvaluated?: number;
  preferredMatches?: number;
  forbiddenMatches?: number;
  missingTranslations?: number;
  exceptionsApplied?: number;
  findings?: number;
}

export interface GlossaryMatch {
  id: string;
  glossaryValidationRunId: string;
  glossaryTermId: string;
  glossaryTranslationId: string | null;
  glossaryVariantId: string | null;
  termCode?: string;
  conceptName?: string;
  languageCode: GlossaryLanguageCode;
  sourceType: string;
  extractedBlockId: string | null;
  ocrBlockId: string | null;
  containerId: string | null;
  detectedSectionId: string | null;
  sectionName?: string | null;
  sourceReference: string;
  matchedText: string;
  normalisedMatchedText: string;
  contextSnippet?: string | null;
  startOffset: number;
  endOffset: number;
  matchType: string;
  isPreferred: boolean;
  isForbidden: boolean;
  exceptionId: string | null;
  metadata: Readonly<Record<string, unknown>>;
  findingId?: string | null;
  findingCode?: string | null;
  findingSeverity?: GlossarySeverity | null;
  createdAt: string;
}

export interface GlossaryMatchListParams {
  page: number;
  pageSize: number;
  search?: string;
  languageCode?: GlossaryLanguageCode;
  termId?: string;
  isPreferred?: boolean;
  isForbidden?: boolean;
  hasException?: boolean;
  findingCode?: string;
}

export type GlossaryMatchList = PaginatedData<GlossaryMatch>;

export interface GlossaryValidationFinding {
  id: string;
  findingCode: string;
  severity: GlossarySeverity;
  status: string;
  title: string;
  description: string;
  recommendation: string;
  glossaryTermId: string;
  languageCode?: GlossaryLanguageCode | null;
  sourceReference?: string | null;
  extractedBlockId: string | null;
  ocrBlockId: string | null;
  translationGroupId: string | null;
  exceptionId: string | null;
  metrics: Readonly<Record<string, unknown>>;
  isRepeat: boolean;
  previousFindingId: string | null;
  createdAt: string;
}

export type GlossaryValidationFindingList = PaginatedData<GlossaryValidationFinding>;

export interface GlossaryHistoryParams {
  page: number;
  pageSize: number;
}

export type GlossaryValidationHistory = PaginatedData<GlossaryValidationRun>;
